Feature/resampling

b-ma opened the pull request:
Hey, 

related to #112 
- `Resampler` now lives in its own file in `media`
- `Buffer::resample` does a linear interpolation

The example `cargo run --example resampling` now sounds better with `decode_audio_data`

Between, I have made a rapid test with the `speex_resample` crate, that seems very slow for us I think and I have a strange bug on the right channel which just disappears. 

To be continued... (seems we will have to dig more on the `rubato` side)

orottier commented on src/buffer.rs:
```diff
@@ -184,12 +181,6 @@ impl AudioBuffer {
         &mut self.channels[index]
     }
 
-    /// Modify every channel in the same way
```
For context of this method see #117 
It's fine that this is dropped now

orottier commented on src/buffer.rs:
```diff
@@ -233,13 +224,14 @@ impl AudioBuffer {
         AudioBuffer::from_channels(channels, sample_rate)
     }
 
-    /// Resample to the desired sample rate.
```
Please leave the `///` in place. If you run rustdoc with `--document-private-items`, it's nice to have this available.
Also put the example usage in three slashes. Not sure why this was not the case already.

b-ma replied:
done (there are ignored from `cargo test --doc` though)

orottier commented on src/buffer.rs:
```diff
@@ -249,7 +241,7 @@ impl AudioBuffer {
     //
     // assert_eq!(
```
Should be `assert_float_eq` now

orottier commented on src/media/resampling.rs:
```diff
@@ -0,0 +1,141 @@
+use crate::media::MediaStream;
+use crate::SampleRate;
+
+/// Sample rate converter and buffer chunk splitter.
```
The docs and example usage are now gone

b-ma replied:
done!

orottier reviewed (changes requested):
Great work. I added a few comments to have a look at and will push a failing test case in a minute

b-ma commented on src/lib.rs:
```diff
@@ -137,6 +138,39 @@ impl AtomicF64 {
 }
 
+/// Utility functions for arguments sanity check
+pub(crate) fn assert_is_valid_sample_rate(sample_rate: SampleRate) {
```
I introduced these function, I thought it could be nice to have more of them to harmonize checks and errors messages (thinking about AudioParams timing arguments that should be checked at some point). Let me know if you agree with that, can be perfectly be moved in `buffer.rs`.

I tried to add them in `ConcreteBaseAudioContext::new` but that was crashing many tests, maybe for another PR if you think we should continue in this direction.

orottier replied:
Looks great, let's do it this way indeed.

Few remarks:
- naming: `assert_is_valid_sample_rate` --> `assert_valid_sample_rate` (same for the others)
- could you document the panics in the docblock, so all callers can copy that in their docblock
- I think `8000 to 96000` is needlessly strict. We only need to guard for `>0` to prevent crashes, so let's do only that?

b-ma replied:
ok done, I have added a comment about "real-life" sample rates in the doc block

b-ma commented:
Hey, just added other checks, example and documentation.

As a side note, we have a discrepancy with types here, i.e. `number_channel: usize` and `length: usize`, which should both be `u32` if we strictly follow the spec. Also this is inconsistent with other places of the code (e.g. `ConcreteBaseAudioContext`). I started to update that but it has many implications here and there so I just reverted back.

Not sure what we should decide though: `unsigned long` is `usize` or `u32`? As they are generally used to as indexes, maybe `usize` would make more sens somehow but I'm not really sure of all the implications of such choice. My only opinion here is that we should be very consistent and apply the same rule everywhere. Maybe creating and maintaining a map of the correspondances somewhere would be helpful? (maybe we have a similar problem with `&[f32]` and `Vec<f32>` at some places)

orottier commented on src/lib.rs:
```diff
@@ -137,6 +138,48 @@ impl AtomicF64 {
+pub(crate) fn assert_is_valid_number_of_channels(number_of_channels: usize) {
+    if number_of_channels == 0 || number_of_channels > MAX_CHANNELS {
+        panic!(
+            "NotSupportedError - Invalid number of channels: {:?} is outside range [1, 32]",
```
Can you replace the harcoded `32` with `{}` and a `MAX_CHANNELS` argument?
In theory, a later version could allow for more channels of course (either by forking and recompiling, or by introducing const generics ourselves)

b-ma replied:
right, it's fixed

orottier commented:
Great work, thanks!
